internal/build/pool.go

```diff
@@ -0,0 +1,112 @@
+package build
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"sync"
+
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/client"
+	digest "github.com/opencontainers/go-digest"
+	"github.com/windmilleng/tilt/internal/model"
+)
+
+var sleepForeverCmd = model.Cmd{Argv: []string{"tail", "-f", "/dev/null"}}
+
+type containerPool struct {
+	docker  *client.Client
+	claimed map[containerID]digest.Digest
+	mu      *sync.Mutex
+}
+
+func newContainerPool(docker *client.Client) containerPool {
+	return containerPool{
+		docker:  docker,
+		claimed: make(map[containerID]digest.Digest, 0),
+		mu:      &sync.Mutex{},
+	}
+}
+
+// Claim a container, creating a new one if necessary.
+//
+// Once the container is claimed, no one is allowed to use it until it's released.
+//
+// TODO(nick): Right now, this always creates a new container.
+func (p containerPool) claimContainer(ctx context.Context, digest digest.Digest) (containerID, error) {
+	config := containerConfigRunCmd(digest, sleepForeverCmd)
+	resp, err := p.docker.ContainerCreate(ctx, config, nil, nil, "")
+	if err != nil {
+		return "", err
+	}
+
+	cID := resp.ID
+
+	err = p.docker.ContainerStart(ctx, cID, types.ContainerStartOptions{})
+	if err != nil {
+		return "", err
+	}
+	containerID := containerID(cID)
+	p.mu.Lock()
+	p.claimed[containerID] = digest
+	p.mu.Unlock()
+	return containerID, nil
+}
+
+// Snapshot a container and release it back into the pool, for
+// use in future builds.
+//
+// TODO(nick): Right now, this deletes the container immediately. We'll eventually
+// add persistence of running containers.
+func (p containerPool) commitContainer(ctx context.Context, cID containerID, entrypoint model.Cmd) (digest.Digest, error) {
+	p.mu.Lock()
+	baseDigest, ok := p.claimed[cID]
+	delete(p.claimed, cID)
+	p.mu.Unlock()
+	if !ok {
+		return "", fmt.Errorf("container %s unknown to pool", baseDigest)
+	}
+
+	opts := types.ContainerCommitOptions{}
+	if !entrypoint.Empty() {
+		// Attach an entrypoint provided by the caller.
+		opts.Changes = []string{entrypoint.EntrypointStr()}
+	} else {
+		// When the container was claimed, we overwrote the entrypoint with tail -f /dev/null.
```

Comment:
teeeechnically we overwrote `cmd` and  _un-set_ the entrypoint (http://bit.ly/2L92LuA). probs most accurate to say: "we overwrote the CMD and the ENTRYPOINT" or something.

Reply:
done